WormsApp/board.cpp

```diff
@@ -1,26 +1,54 @@
 #include "board.h"
+#include <iostream>
+#include <thread>
+#include <unordered_map>
 
 
 Board::Board(int width, int height) : nextId_{1}, killAll_{false} {
-  board_.resize(height);
-  for (int i = 0; i < (int)board_.size(); ++i)
+
+    if (width <= 0 || height <= 0)
+        throw std::invalid_argument("Width and Height parameters should be grater than 0.");
+    board_.resize(height);
+    for (int i = 0; i < (int)board_.size(); ++i)
     board_[i].resize(width);
 }
 
+Board::~Board() {
+ for(auto& w: getWorms()){
+     w.second.join();}
+ }
+
+
 void Board::addWorm(WormType type, int x, int y) {
-  // TODO: Implement the body of the function. It should add a worm
+  // TODO:   the body of the function. It should add a worm
   // into the board_ and start a new thread which will invoke
   // call operator - operator()(int) - defined in Worm class.
   // Note: there are a few class members (like wormTypes_) that need to be
   // updated as well to have full information about the worm that is being
   // added. They are usually based on worm's id which should be uniquely given
   // here.
+     board_[x][y] = nextId_;
+    if (type == Lazy)
+    {
+        LazyWorm w = LazyWorm(x,y,this);
+        worms_.add(nextId_,std::thread(std::move(w),nextId_));
```

Comment:
I recently discovered that this creates and instance of a `std::thread` and then copies it when inserting to the map. In general this may be dangerous depending on the implementation of a copy constructor and sometimes may not compile (when copy constructor is not defined). Fortunately thread is implemented wisely and will take care of it - copy constructor as well as assignment operator are moving the thread to the new one. Inside, new thread is created and immediately joined with the old one.

TLDR: It's ok, you just need to be careful in the future.

Reply:
Ok. I see)